### PR TITLE
DEVPROD-57: Fetch logs using the log fetching task methods

### PR DESCRIPTION
**Pull request by julianedwards**
[DEVPROD-57](https://jira.mongodb.org/browse/DEVPROD-57)

### Description
Fetch logs using the log fetching task methods.

For the sake of readable PRs and safer deploys, I decided to break this work up. The first part (this PR) temporarily adds fallback code to support DB logs in the task output interface, a subsequent PR will remove support for DB logs, update the agent code, and migrate the local and smoke data to use a GridFS backed pail bucket for log data.

- Move task log model into its own package `model/tasklog` to avoid import cycle in `taskoutput`
- Move test log model into its own package `model/testlog` to avoid import cycle in `taskoutput`
- Create temporary log iterator for the `apimodels.LogMessage` type
- Add temporary DB fallback logic for fetching task and test log fetching in the `taskoutput` package
- Replace all log fetching code with the new log fetching task models in the `service` and `graphql` packages
- Remove buildlogger-specific logic from legacy UI code

### Testing
Tested in staging that we can still write and fetch task and test logs.

### Documentation
N/A

### CODE REVIEW

File: apimodels/task_log.go [hunk 1]
```diff
@@ -54,149 +45,121 @@ type TaskLog struct {
...
-type GetBuildloggerLogsOptions struct {
-	BaseURL       string `json:"_"`
-	TaskID        string `json:"-"`
-	TestName      string `json:"-"`
-	GroupID       string `json:"-"`
-	Execution     *int   `json:"-"`
-	PrintPriority bool   `json:"-"`
-	Tail          int    `json:"-"`
-	LogType       string `json:"-"`
+func GetPriority(prefix string) level.Priority {
```
Comment (ablack12) [inline apimodels/task_log.go:63]: can you add a function comment here (unless this is mega temporary)

File: apimodels/task_log.go [hunk 2]
```diff
@@ -54,149 +45,121 @@ type TaskLog struct {
...
 
-// ReadBuildloggerToChan parses Cedar buildlogger log lines by message and
-// severity and reads into a channel.
-func ReadBuildloggerToChan(ctx context.Context, taskID string, r io.ReadCloser, lines chan<- LogMessage) {
-	var (
-		line string
-		err  error
-	)
+// StreamFromLogIterator streams log lines from the given iterator to the
+// returned log message channel. It is the responsibility to close the log
```
Comment (ablack12) [inline apimodels/task_log.go:95]: "It is the responsibility to close the log" -- can you fix this wording?

File: graphql/util.go [hunk 3]
```diff
@@ -915,6 +917,35 @@ func getProjectMetadata(ctx context.Context, projectId *string, patchId *string)
...
 }
 
+//////////////////////////////////
+// Helper functions for task logs.
+//////////////////////////////////
+
+func getTaskLogs(ctx context.Context, obj *TaskLogs, logType taskoutput.TaskLogType) ([]*apimodels.LogMessage, error) {
+	dbTask, err := task.FindOneIdAndExecution(obj.TaskID, obj.Execution)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("Finding task %s: %s", obj.TaskID, err.Error()))
```
Comment (ablack12) [inline graphql/util.go:927]: task '%s'

File: model/log/service_v0.go [hunk 4]
```diff
@@ -113,7 +113,7 @@ func (s *logServiceV0) getLogChunks(ctx context.Context, logNames []string) (map
 
 	for _, chunks := range logChunks {
 		sort.Slice(chunks, func(i, j int) bool {
-			return chunks[i].start > chunks[j].start
+			return chunks[i].start < chunks[j].start
```
Comment (ablack12) [inline model/log/service_v0.go:116]: what's the thought behind this?

**Review by ablack12 — APPROVED** [review]

File: model/log/service_v0.go (hunk 4 above)
Comment (julianedwards) [reply model/log/service_v0.go:116]: We want to iterate through chunks of log lines in ascending order rather than descending order.

File: apimodels/task_log.go (hunk 1 above)
Comment (julianedwards) [reply apimodels/task_log.go:63]: It doesn't need to be exported so I just made it private, it will also be temporary.

File: apimodels/task_log.go (hunk 2 above)
Comment (julianedwards) [reply apimodels/task_log.go:95]: Done.

File: graphql/util.go (hunk 3 above)
Comment (julianedwards) [reply graphql/util.go:927]: Good catch! Done.

File: service/ui.go [hunk 5]
```diff
@@ -319,7 +319,6 @@ func (uis *UIServer) GetServiceApp() *gimlet.APIApp {
 	// Proxy downloads for task uploaded files via S3
 	app.AddRoute(("/task_file_raw/{task_id}/{execution}/{file_name}")).Wrap(needsLogin, needsContext, allowsCORS, viewLogs).Handler(uis.taskFileRaw).Get()
 	// Test Logs
-	app.AddRoute("/test_log/{log_id}").Wrap(needsLogin, needsContext, allowsCORS).Handler(uis.testLog).Get()
```
Comment (khelif96) [inline service/ui.go:322]: Are we sure this route is no longer in use?

File: service/task.go [hunk 6]
```diff
@@ -507,140 +493,122 @@ func getTaskDependencies(t *task.Task) ([]uiDep, string, error) {
...
 func (uis *UIServer) taskLogRaw(w http.ResponseWriter, r *http.Request) {
 	projCtx := MustHaveProjectContext(r)
 	if projCtx.Task == nil {
-		http.Error(w, "Not found", http.StatusNotFound)
+		http.Error(w, "task not found", http.StatusNotFound)
 		return
 	}
 
-	raw := (r.FormValue("text") == "true") || (r.Header.Get("Content-Type") == "text/plain")
-	printPriority := r.FormValue("priority") == "true" || !raw
```
Comment (khelif96) [inline service/task.go:584]: Looks like this is missing priority information if the `priority=true` is set.

Before:
image

After:
image

Comment (khelif96) [reply service/task.go:584]: Parsley passes the following query params to request task logs.
`?priority=true&text=true&type=T`
It looks like priority might be getting handled correctly.

**Review by khelif96 — CHANGES_REQUESTED** [review]

File: graphql/tests/task/taskLogs/results.json [hunk 7]
```diff
@@ -8,11 +8,11 @@
             "taskLogs": {
               "systemLogs": [
                 {
-                  "type": "S",
+                  "type": "",
                   "severity": "I",
                   "message": "    [javac]     protected void finalize() {",
                   "timestamp": "2019-12-07T13:31:19.393-05:00",
-                  "version": 1
+                  "version": 0
```
Comment (khelif96) [inline graphql/tests/task/taskLogs/results.json:15]: What is the motivation behind these changes?

Comment (julianedwards) [reply graphql/tests/task/taskLogs/results.json:15]: That was the DB version of the logs, which we no longer support.

File: service/task.go (hunk 6 above)
Comment (julianedwards) [reply service/task.go:584]: Fixed!
Screenshot 2023-11-01 at 11 48 06 AM

**Comment by khelif96** [comment]
The Parsley e2e tests are now failing it looks like the timestamps are no longer formatted the same way In the log file and may even be in a different timezone. @julianedwards
